sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -830,11 +845,184 @@ public void populateDisplayData(DisplayData.Builder builder) {
       inner.populateDisplayData(builder);
     }
 
+    private boolean hasStatementAndSetter() {
+      return inner.getStatement() != null && inner.getPreparedStatementSetter() != null;
+    }
+
     @Override
     public PDone expand(PCollection<T> input) {
+      // fixme: validate invalid table input
+      if (input.hasSchema() && !hasStatementAndSetter()) {
+        checkArgument(inner.getTable() != null, "table cannot be null if statement is not provided");
+        Schema schema = input.getSchema();
+        List<Schema.Field> fields = getFilteredFields(schema);
+        inner = inner.withStatement(JdbcUtil.generateStatement(inner.getTable(), fields));
+        inner = inner.withPreparedStatementSetter(new AutoGeneratedPreparedStatementSetter(fields, input.getToRowFunction()));
+      }
+
       inner.expand(input);
       return PDone.in(input.getPipeline());
     }
+
+    private List<Schema.Field> getFilteredFields(Schema schema) {
+      List<Schema.Field> fields = schema.getFields();
+      Connection connection;
+      try {
+        connection = inner.getDataSourceProviderFn().apply(null).getConnection();
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable());
+      }
+      List<String> databaseColumnList = getTableColumnList(connection);
+      fields = fields.stream().filter((field) -> {
+        return databaseColumnList.contains(field.getName().toLowerCase());
+      }).collect(Collectors.toList());
+      return fields;
+    }
+
+    private List<String> getTableColumnList(Connection connection) {
+      try {
+        List<String> tableColumns = new ArrayList<>();
+        // get columns of database table
+        ResultSet resultSet = connection.getMetaData().getColumns(null, null, inner.getTable(), null);
+        while (resultSet.next()) {
+          tableColumns.add(resultSet.getString("column_name").toLowerCase());
+        }
+        return tableColumns;
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable());
+      }
+    }
+
+    /**
+     * Interface implemented by functions that sets prepared statement data
+     */
+    @FunctionalInterface
+    interface PreparedStatementSetCaller extends Serializable {
+      void set(Row element, PreparedStatement preparedStatement, int index) throws SQLException;
+    }
+
+    private class AutoGeneratedPreparedStatementSetter implements PreparedStatementSetter<T> {
+
+      private List<Schema.Field> fields;
+      private SerializableFunction<T, Row> toRowFn;
+      Map<Integer, PreparedStatementSetCaller> preparedStatementFieldSetterMap = new HashMap<>();
```

Comment:
Why a map indexed by an integer when an array would do?

Schema field order is immutable. So you can just create a list of PreparedStatementSetCaller objects and then iterate through it.

Reply:
Done